### PR TITLE
[MI-2597]:Fixed github issue mattermost#500 for emoji synchronisation

### PR DESCRIPTION
**Pull request by Kshitij-Katiyar**
https://github.com/mattermost/mattermost-plugin-github/issues/500

### CODE REVIEW

File: server/plugin/plugin.go [hunk 1]
```diff
@@ -111,6 +113,29 @@ func NewPlugin() *Plugin {
...
+
+	var githubEmojiMap = map[string]string{}
+	for systemEmoji := range model.SystemEmojis {
+		for mmBase, ghBase := range baseGithubEmojiMap {
+			if strings.HasPrefix(systemEmoji, mmBase) {
+				githubEmojiMap[systemEmoji] = ghBase
+			}
+		}
+	}
+	p.emojiMap = githubEmojiMap
```
Comment (ayusht2810) [inline server/plugin/plugin.go]:
```suggestion
	}
	
	p.emojiMap = githubEmojiMap
```

File: server/plugin/plugin.go [hunk 2]
```diff
@@ -111,6 +113,29 @@ func NewPlugin() *Plugin {
...
+
```
Comment (ayusht2810) [inline server/plugin/plugin.go:137]:
```suggestion
```

File: server/plugin/plugin.go [hunk 3]
```diff
@@ -250,6 +275,151 @@ func (p *Plugin) OnDeactivate() error {
...
+	}
+
+	repo, ok := post.GetProp(postPropGithubRepo).(string)
+	if !ok || repo == "" {
+		return orgRepo, id, objectType, false
+	}
+
+	orgRepo = strings.Split(repo, "/")
+	if len(orgRepo) != 2 {
+		p.API.LogError("Invalid orgainsation repository")
```
Comment (ayusht2810) [inline server/plugin/plugin.go]:
```suggestion
		p.API.LogError("Invalid organisation repository")
```

File: server/plugin/plugin.go [hunk 4]
```diff
@@ -250,6 +275,151 @@ func (p *Plugin) OnDeactivate() error {
...
+	if appErr != nil {
+		if appErr.ID != apiErrorIDNotConnected {
+			p.API.LogError("Error in getting user info", "error", appErr.Message)
+		}
+		return
+	}
+	ghClient := p.githubConnectUser(context.Background(), info)
+
+	switch objectType {
+	case githubObjectTypeIssueComment:
```
Comment (ayusht2810) [inline server/plugin/plugin.go]:
```suggestion
	}
	
	ghClient := p.githubConnectUser(context.Background(), info)
	switch objectType {
	case githubObjectTypeIssueComment:
```

File: server/plugin/plugin.go [hunk 5]
```diff
@@ -250,6 +275,151 @@ func (p *Plugin) OnDeactivate() error {
...
+		if _, _, err := ghClient.Reactions.CreateIssueCommentReaction(context.Background(), orgRepo[0], orgRepo[1], int64(id), githubEmoji); err != nil {
+			p.API.LogDebug("Error found while creating issue comment reaction", err)
+			return
+		}
+	case githubObjectTypeIssue:
+		if _, _, err := ghClient.Reactions.CreateIssueReaction(context.Background(), orgRepo[0], orgRepo[1], int(id), githubEmoji); err != nil {
+			p.API.LogDebug("Error found while creating issue reaction", err)
+			return
+		}
+
```
Comment (ayusht2810) [inline server/plugin/plugin.go]:
```suggestion
```

File: server/plugin/plugin.go [hunk 6]
```diff
@@ -250,6 +275,151 @@ func (p *Plugin) OnDeactivate() error {
...
+	case githubObjectTypePRReviewComment:
+		if _, _, err := ghClient.Reactions.CreatePullRequestCommentReaction(context.Background(), orgRepo[0], orgRepo[1], int64(id), githubEmoji); err != nil {
+			p.API.LogDebug("Error found while creating pr review comment reaction", err)
```
Comment (ayusht2810) [inline server/plugin/plugin.go]:
```suggestion
			p.API.LogDebug("Error found while creating PR review comment reaction", err)
```

Comment (ayusht2810) [reply server/plugin/plugin.go]: maybe

File: server/plugin/plugin.go [hunk 7]
```diff
@@ -250,6 +275,151 @@ func (p *Plugin) OnDeactivate() error {
...
+	if appErr != nil {
+		if appErr.ID != apiErrorIDNotConnected {
+			p.API.LogError("Error in getting user info", "error", appErr.Message)
+		}
+		return
+	}
+	ghClient := p.githubConnectUser(context.Background(), info)
+
+	switch objectType {
+	case githubObjectTypeIssueComment:
```
Comment (ayusht2810) [inline server/plugin/plugin.go]: same

File: server/plugin/plugin.go [hunk 8]
```diff
@@ -250,6 +275,151 @@ func (p *Plugin) OnDeactivate() error {
...
+
+		for _, reactionObj := range reactions {
+			if info.UserID == reaction.UserId && p.emojiMap[reaction.EmojiName] == reactionObj.GetContent() {
+				if _, err = ghClient.Reactions.DeleteIssueReaction(context.Background(), orgRepo[0], orgRepo[1], int(id), reactionObj.GetID()); err != nil {
+					p.API.LogDebug("Error found while removing issue reaction", "error", err.Error())
+				}
+				return
+			}
+		}
+
```
Comment (ayusht2810) [inline server/plugin/plugin.go]:
```suggestion
```

File: server/plugin/plugin.go [hunk 9]
```diff
@@ -250,6 +275,151 @@ func (p *Plugin) OnDeactivate() error {
...
+	case githubObjectTypePRReviewComment:
+		reactions, _, err := ghClient.Reactions.ListPullRequestCommentReactions(context.Background(), orgRepo[0], orgRepo[1], int64(id), &github.ListOptions{})
+		if err != nil {
+			p.API.LogDebug("Error getting pr review comment reaction list", err)
```
Comment (ayusht2810) [inline server/plugin/plugin.go]: same

Comment (ayusht2810) [reply server/plugin/plugin.go]: check everywhere, if changing

**Review by ayusht2810 — APPROVED** [review]

File: server/plugin/plugin.go [hunk 10]
```diff
@@ -250,6 +275,149 @@ func (p *Plugin) OnDeactivate() error {
 	return nil
 }
 
+func (p *Plugin) getPostPropsForReaction(reaction *model.Reaction) (orgRepo []string, id float64, objectType string, ok bool) {
+	post, err := p.client.Post.GetPost(reaction.PostId)
+	if err != nil {
+		p.API.LogError("Error fetching post for reaction", "error", err.Error())
```
Comment (raghavaggarwal2308) [inline server/plugin/plugin.go]:
```suggestion
		p.API.LogError("Error fetching post for reaction", "Error", err.Error())
```
check everywhere. Can skip this if we need to keep it in sync with existing code

File: server/plugin/plugin.go [hunk 11]
```diff
@@ -250,6 +275,149 @@ func (p *Plugin) OnDeactivate() error {
...
+		reactions, _, err := ghClient.Reactions.ListIssueCommentReactions(context.Background(), orgRepo[0], orgRepo[1], int64(id), &github.ListOptions{})
+		if err != nil {
+			p.API.LogDebug("Error getting issue comment reaction list", err)
+			return
+		}
+
+		for _, reactionObj := range reactions {
+			if info.UserID == reaction.UserId && p.emojiMap[reaction.EmojiName] == reactionObj.GetContent() {
+				if _, err = ghClient.Reactions.DeleteIssueCommentReaction(context.Background(), orgRepo[0], orgRepo[1], int64(id), reactionObj.GetID()); err != nil {
+					p.API.LogDebug("Error found while removing issue comment reaction", "error", err.Error())
```
Comment (raghavaggarwal2308) [inline server/plugin/plugin.go]: why `LogDebug?`

**Review by raghavaggarwal2308 — CHANGES_REQUESTED** [review]

File: server/plugin/plugin.go (hunk 11 above)
Comment (Kshitij-Katiyar) [reply server/plugin/plugin.go]: Its given by mattermost that they wanted this as LogDebug

File: server/plugin/plugin.go (hunk 10 above)
Comment (Kshitij-Katiyar) [reply server/plugin/plugin.go]: keep it in sync with existing code

File: server/plugin/plugin.go [hunk 12]
```diff
@@ -111,6 +113,29 @@ func NewPlugin() *Plugin {
...
+		"heart":      "heart",
+		"tada":       "hooray",
+		"rocket":     "rocket",
+		"eyes":       "eyes",
+	}
+
+	var githubEmojiMap = map[string]string{}
+	for systemEmoji := range model.SystemEmojis {
+		for mmBase, ghBase := range baseGithubEmojiMap {
+			if strings.HasPrefix(systemEmoji, mmBase) {
```
Comment (manojmalik20) [inline server/plugin/plugin.go:132]: What was the need for iterating on `model.SystemEmojis` map? Was it for supporting emojis with different skin tones?

File: server/plugin/plugin.go [hunk 13]
```diff
@@ -111,6 +113,29 @@ func NewPlugin() *Plugin {
 		"issue":         p.handleIssue,
 	}
 
+	var baseGithubEmojiMap = map[string]string{
```
Comment (manojmalik20) [inline server/plugin/plugin.go]:
```suggestion
	baseGithubEmojiMap := map[string]string{
```

File: server/plugin/plugin.go [hunk 14]
```diff
@@ -111,6 +113,29 @@ func NewPlugin() *Plugin {
...
+		"thumbsdown": "-1",
+		"laughing":   "laugh",
+		"confused":   "confused",
+		"heart":      "heart",
+		"tada":       "hooray",
+		"rocket":     "rocket",
+		"eyes":       "eyes",
+	}
+
+	var githubEmojiMap = map[string]string{}
```
Comment (manojmalik20) [inline server/plugin/plugin.go]:
```suggestion
	p.emojiMap = map[string]string{}
```

File: server/plugin/plugin.go [hunk 15]
```diff
@@ -111,6 +113,29 @@ func NewPlugin() *Plugin {
...
+	for systemEmoji := range model.SystemEmojis {
+		for mmBase, ghBase := range baseGithubEmojiMap {
+			if strings.HasPrefix(systemEmoji, mmBase) {
+				githubEmojiMap[systemEmoji] = ghBase
+			}
+		}
+	}
+
+	p.emojiMap = githubEmojiMap
```
Comment (manojmalik20) [inline server/plugin/plugin.go]:
```suggestion
```

File: server/plugin/plugin.go [hunk 16]
```diff
@@ -250,6 +275,149 @@ func (p *Plugin) OnDeactivate() error {
...
+
+	orgRepo, id, objectType, ok := p.getPostPropsForReaction(reaction)
+	if !ok {
+		return
+	}
+
+	info, appErr := p.getGitHubUserInfo(reaction.UserId)
+	if appErr != nil {
+		if appErr.ID != apiErrorIDNotConnected {
+			p.API.LogError("Error in getting user info", "error", appErr.Message)
```
Comment (manojmalik20) [inline server/plugin/plugin.go]:
```suggestion
			p.API.LogDebug("Error in getting user info", "error", appErr.Message)
```

File: server/plugin/plugin.go (hunk 10 above)
Comment (manojmalik20) [reply server/plugin/plugin.go]:
```suggestion
		p.API.LogDebug("Error fetching post for reaction", "error", err.Error())
```

File: server/plugin/plugin.go [hunk 17]
```diff
@@ -250,6 +275,149 @@ func (p *Plugin) OnDeactivate() error {
...
+		}
+		return
+	}
+
+	ghClient := p.githubConnectUser(context.Background(), info)
+	switch objectType {
+	case githubObjectTypeIssueComment:
+		if _, _, err := ghClient.Reactions.CreateIssueCommentReaction(context.Background(), orgRepo[0], orgRepo[1], int64(id), githubEmoji); err != nil {
+			p.API.LogDebug("Error found while creating issue comment reaction", err)
```
Comment (manojmalik20) [inline server/plugin/plugin.go]:
```suggestion
			p.API.LogDebug("Error occurred while creating issue comment reaction", "error", err.Error())
```

File: server/plugin/plugin.go (hunk 11 above)
Comment (manojmalik20) [inline server/plugin/plugin.go]:
```suggestion
					p.API.LogDebug("Error occurred while removing issue comment reaction", "error", err.Error())
```
I think `found` is the wrong word to use. Fix everywhere.

File: server/plugin/plugin.go [hunk 18]
```diff
@@ -250,6 +275,149 @@ func (p *Plugin) OnDeactivate() error {
...
+				if _, err = ghClient.Reactions.DeleteIssueReaction(context.Background(), orgRepo[0], orgRepo[1], int(id), reactionObj.GetID()); err != nil {
+					p.API.LogDebug("Error found while removing issue reaction", "error", err.Error())
+				}
+				return
+			}
+		}
+	case githubObjectTypePRReviewComment:
+		reactions, _, err := ghClient.Reactions.ListPullRequestCommentReactions(context.Background(), orgRepo[0], orgRepo[1], int64(id), &github.ListOptions{})
+		if err != nil {
+			p.API.LogDebug("Error getting PR review comment reaction list", err)
```
Comment (manojmalik20) [inline server/plugin/plugin.go]:
```suggestion
			p.API.LogDebug("Error getting PR review comment reaction list", "error", err.Error())
```
Fix everywhere.

File: server/plugin/plugin.go [hunk 19]
```diff
@@ -250,6 +275,149 @@ func (p *Plugin) OnDeactivate() error {
...
+	info, appErr := p.getGitHubUserInfo(reaction.UserId)
+	if appErr != nil {
+		if appErr.ID != apiErrorIDNotConnected {
+			p.API.LogError("Error in getting user info", "error", appErr.Message)
+		}
+		return
+	}
+
+	ghClient := p.githubConnectUser(context.Background(), info)
+	switch objectType {
```
Comment (manojmalik20) [inline server/plugin/plugin.go:329]:
```suggestion
        var1, var2 := orgRepo[0], orgRepo[1]
	switch objectType {
```

File: server/plugin/plugin.go (hunk 12 above)
Comment (Kshitij-Katiyar) [reply server/plugin/plugin.go:132]: yes manoj

File: server/plugin/plugin.go [hunk 20]
```diff
@@ -320,26 +319,27 @@ func (p *Plugin) ReactionHasBeenAdded(c *plugin.Context, reaction *model.Reactio
...
 	}
 
 	ghClient := p.githubConnectUser(context.Background(), info)
+	owner, repo := orgRepo[0], orgRepo[1]
 	switch objectType {
 	case githubObjectTypeIssueComment:
-		if _, _, err := ghClient.Reactions.CreateIssueCommentReaction(context.Background(), orgRepo[0], orgRepo[1], int64(id), githubEmoji); err != nil {
-			p.API.LogDebug("Error found while creating issue comment reaction", err)
+		if _, _, err := ghClient.Reactions.CreateIssueCommentReaction(context.Background(), owner, repo, int64(id), githubEmoji); err != nil {
+			p.API.LogDebug("Error found occurred creating issue comment reaction", "error", err.Error())
```
Comment (manojmalik20) [inline server/plugin/plugin.go]:
```suggestion
			p.API.LogDebug("Error occurred while creating issue comment reaction", "error", err.Error())
```
Fix everywhere.

**Review by manojmalik20 — COMMENTED** [review]
Please fix the lint errors as well.

**Review by manojmalik20 — APPROVED** [review]

**Pull request merged**